sourcemap/elasticsearch.go

```diff
@@ -0,0 +1,115 @@
+package sourcemap
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/go-sourcemap/sourcemap"
+
+	"github.com/elastic/beats/libbeat/common"
+	es "github.com/elastic/beats/libbeat/outputs/elasticsearch"
+)
+
+type elasticsearch interface {
+	fetch(id Id) (*sourcemap.Consumer, error)
+}
+
+type smapElasticsearch struct {
+	clients []es.Client
+	index   string
+}
+
+func NewElasticsearch(config *common.Config, index string) (*smapElasticsearch, error) {
+	esClients, err := es.NewElasticsearchClients(config)
+	if err != nil || esClients == nil || len(esClients) == 0 {
+		return nil, Error{Msg: fmt.Sprintf("Sourcemap ES Client cannot be initialized. %v", err.Error()), Kind: InitError}
+	}
+	if index == "" {
+		index = "*"
+	}
+	return &smapElasticsearch{clients: esClients, index: index}, nil
+}
+
+func (e *smapElasticsearch) fetch(id Id) (*sourcemap.Consumer, error) {
+	result, err := e.runESQuery(query(id))
+	if err != nil {
+		return nil, err
+	}
+	return parseResult(result, id)
+}
+
+func (e *smapElasticsearch) runESQuery(body map[string]interface{}) (*es.SearchResults, error) {
+	var err error
+	var result *es.SearchResults
+	for _, client := range e.clients {
+		_, result, err = client.Connection.SearchURIWithBody(e.index, "", nil, body)
+		if err == nil {
+			return result, nil
+		}
+	}
+	if err != nil {
+		return nil, Error{Msg: err.Error(), Kind: AccessError}
+	}
+	return result, nil
```

Comment:
@simitt this line never will be reached, right? Also, if the program reaches L50, `err` will always be `not nil` because setup ensures that there is at least 1 client.

Reply:
hm, yes I think you're right.